File: llm/lib/tasks/deep-research.mjs
// deep-research — multi-round retrieval + synthesis, streaming NDJSON.
import { createNdjsonWriter } from '../stream.mjs';
import { runPrompt } from '../runner.mjs';
import { runResearchLoop } from '../research.mjs';

export async function run({ kbRoot, input, outputPath }) {
  const writer = createNdjsonWriter(outputPath);
  const question = input?.question || '';
  const opts = input?.opts || {};

  let context = '';
  let citations = [];
  try {
    const result = await runResearchLoop({
      kbRoot,
      question,
      onEvent: (e) => writer.write(e),
      opts,
    });
    context = result.context;
    citations = result.citations;
  } catch (err) {
    writer.write({ type: 'error', message: err.message });
  }

  try {
    await runPrompt(kbRoot, 'deep-research', { question, context }, {
      stream: true,
      onDelta: (delta) => writer.write({ type: 'chunk', text: delta }),
    });
  } catch (err) {
    writer.write({ type: 'error', message: err.message });
  }

  writer.write({ type: 'done', citations });
  writer.end();
  await writer.finish();

  return { rounds: opts.maxRounds || 3, tokens_in: 0, tokens_out: 0 };
}

File: agent/weft_agent/tasks/deep_research.py
"""deep-research — multi-round retrieval + synthesis, streaming NDJSON
(port of llm/lib/tasks/deep-research.mjs).
"""
from ..research import run_research_loop
from ..runner import run_prompt
from ..stream import NdjsonWriter


def run(kb_root, input=None, output_path=None):
    input = input or {}
    writer = NdjsonWriter(output_path)
    question = input.get("question") or ""
    opts = input.get("opts") or {}

    context = ""
    citations: list[str] = []
    try:
        result = run_research_loop(kb_root, question, on_event=writer.write, opts=opts)
        context = result["context"]
        citations = result["citations"]
    except Exception as err:  # noqa: BLE001
        writer.write({"type": "error", "message": str(err)})

    try:
        run_prompt(kb_root, "deep-research", {"question": question, "context": context},
                   stream=True, on_delta=lambda d: writer.write({"type": "chunk", "text": d}))
    except Exception as err:  # noqa: BLE001
        writer.write({"type": "error", "message": str(err)})

    writer.write({"type": "done", "citations": citations})
    writer.end()

    return {"rounds": opts.get("maxRounds") or 3, "tokens_in": 0, "tokens_out": 0}
